src/tree.rs

```diff
@@ -3,79 +3,270 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::{cmp::{Eq, PartialEq},
-          collections::HashSet,
+          collections::{HashMap, HashSet},
           fmt,
           hash::{Hash, Hasher},
-          iter};
+          ops::Deref};
 
-/// Synced item kinds. Each corresponds to a Sync record type.
-#[derive(Eq, Hash, PartialEq)]
-pub enum Kind {
-    Bookmark,
-    Query,
-    Folder,
-    Livemark,
-    Separator,
+/// A complete, rooted bookmark tree with tombstones.
+///
+/// The tree stores bookmark nodes in a vector, and uses indices in the vector
+/// to identify parents and children. This makes traversal and lookup very
+/// efficient. Retrieving a node's parent takes one indexing operation,
+/// retrieving children takes one indexing operation per child, and retrieving a
+/// node by random GUID takes one hash map lookup and one indexing operation.
+#[derive(Debug)]
+pub struct Tree {
+    entries: Vec<Entry>,
+    index_by_guid: HashMap<String, usize>,
+    deleted_guids: HashSet<String>,
 }
 
-impl fmt::Display for Kind {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.write_str(match self {
-                        Kind::Bookmark => "Bookmark",
-                        Kind::Query => "Query",
-                        Kind::Folder => "Folder",
-                        Kind::Livemark => "Livemark",
-                        Kind::Separator => "Separator",
-                    })
+impl Tree {
+    /// Constructs a new rooted tree.
+    pub fn new(root: Item) -> Tree {
+        let mut index_by_guid = HashMap::new();
+        index_by_guid.insert(root.guid.to_string(), 0);
+        Tree { entries: vec![Entry { parent_index: None,
+                                     item: root,
+                                     level: 0,
+                                     child_indices: Vec::new(), }],
+               index_by_guid,
+               deleted_guids: HashSet::new(), }
+    }
+
+    pub fn deletions(&self) -> Vec<&str> {
+        self.deleted_guids
+            .iter()
+            .map(|guid| guid.as_ref())
+            .collect()
+    }
+
+    pub fn is_deleted(&self, guid: &str) -> bool {
+        self.deleted_guids.contains(guid)
+    }
+
+    pub fn note_deleted<T>(&mut self, guid: T)
+        where T: Into<String> {
+        self.deleted_guids.insert(guid.into());
+    }
+
+    pub fn guids(&self) -> Vec<&str> {
+        self.index_by_guid
+            .keys()
+            .map(|guid| guid.as_ref())
+            .chain(self.deleted_guids.iter().map(|guid| guid.as_ref()))
+            .collect()
+    }
+
+    pub fn node_for_guid<T>(&self, guid: T) -> Option<Node>
+        where T: AsRef<str> {
+        self.index_by_guid
+            .get(guid.as_ref())
+            .map(|index| self.node(*index))
+    }
+
+    pub fn insert<T>(&mut self, parent_guid: T, item: Item)
+        where T: AsRef<str> {
+        if self.index_by_guid.contains_key(&item.guid) {
```

Comment:
Nit: This should just be an assert probably (you can still provide a string with a format)